[STONEINTG-317] Git repo support for the Integration Test Scenarios (#92)
redhat-appstudio/book: ADR/0023-git-references-to-furnish-integration-test-scenarios.md

to allow using different types of Tekton resolvers ex: bundles-resolver, cluster-resolver, hub-resolver along with 
the git-resolver. This gives the extensibility to permit the use of different resolvers in the future.

The new format of adding the Integration Test Scenarios would look like:

![](../diagrams/ADR-0023/git-references-ITS.jpg)

Users git repository path to their test code can now provide the information or the location of their IntegrationTestScenarios
through Git references like **Github URL**, **Git Options** (**Revision** & **Path in repository**). 

The information provided by the users in the above form will then be consumed by the integration service with the help of IntegrationTestScenario CR and will be used to run the integration tests against the Application Snapshot in question.

Example of the tekton pipeline definition with git resolver:

![](../diagrams/ADR-0023/tekton-pipeline-definition-git-resolver.jpg)

## Consequences

* Better user experience through Tekton pipeline definitions that are hosted in git repositories which are easier to 

> maintain as opposed to build and maintain the Tekton bundle images.

Reviewer: ```suggestion
maintain as opposed to building and maintaining the Tekton bundle images.
```

Reviewer 2: ```suggestion
maintain as opposed to building their own Tekton bundle images every time they adjust their test. It also removes the need for our users to have to understand what a Tekton bundle is and learn how to build it, if they don't already know.
```

What you had was fair, but I think undersells the improvement this provides for our users. Not sure how in depth the ADR needs to go though. I also dropped the "and maintaining" in my suggestion since it already mentioned that it was easier to maintain earlier in the sentence.

Author: ack

* Resolvers are currently in Technology Preview for the OpenShift Pipelines 1.9. Technology Preview features are not
supported with Red Hat production service level agreements (SLAs) and might not be functionally complete.
Red Hat does not recommend using them in production. These features only provide early access to upcoming product features,
enabling customers to test functionality and provide feedback during the development process.

## Footnotes

* The risk mentioned in the consequences has been [accepted] by the PM(s).
* With the current design, there will be no direct support for the Tekton bundles. 

>   It could only be supported through the tekton bundle resolver.

Reviewer 2: Does the ADR need to mention anything about migrating existing scenarios to the new format (STONEINTG-356)?

Author: Addressed at line [#48](https://github.com/redhat-appstudio/book/pull/92/files#diff-4adb99d2e0cdfd21098edf4375e38cd0ca48caeb8f285113121a39634a3989abR48)

Reviewer 2: That looks good to me, thank you!

# 23. Git references to furnish Integration Test Scenarios

* Date Documented: 2023-04-10
* Date Accepted: TBD

## Status

In Review

## Context

> So far Integration service only supports setting the Tekton bundle (Bundle string `json:"bundle"`) in the 

Reviewer: ```suggestion
Up to now, the Integration service has only supported setting the Tekton bundle (Bundle string `json:"bundle"`) in the 
```

IntegrationTestScenario [CR] as a reference for the integration tests, in order to run the Tekton PipelineRuns.

Below is the existing format to supply the Integration test references:

![](../diagrams/ADR-0023/tekton-bundle-ITS.jpg)

> Though building and maintaining the Tekton bundles would be cumbersome for the users when providing the custom tests.

Reviewer: ```suggestion
Building and maintaining Tekton bundles is cumbersome for users when providing  custom tests.
```

Resolving tasks from a git repository is more natural for developers and allows for faster iteration. Hence, users

> should be able to reference through Tekton pipeline definitions that are hosted in git repositories ([RHTAP-402]).

Reviewer: ```suggestion
should be able to reference their Tekton pipeline definitions that are hosted in git repositories ([RHTAP-402]).
```

Author: ack

To support the functionality of using git references, Tekton resolvers need to be enabled. As part of [PLNSRVCE-1030], 
pipeline service adds support for the [Tekton-resolvers] including git resolvers which should in turn give us the ability
to expand the support for resolvers in the IntegrationTestScenarios.

## Decision

> Integration service now migrates to the new version of CRD having the expanded functionality to allow using different

Reviewer: ```suggestion
Integration service now migrates to the new version of the IntegrationTestScenario CRD having the expanded functionality to allow using different
```

types of Tekton resolvers ex: bundles-resolver, cluster-resolver, hub-resolver along with the git-resolver.
This gives the extensibility to permit the use of different resolvers in the future.

The new format of adding the Integration Test Scenarios would look like:

![](../diagrams/ADR-0023/git-references-ITS.jpg)

> In lieu of Tekton bundles which are burdensome to maintain, users git repository path to their test code can now

Reviewer: This sentence does not read well. unable to make a suggestion though

provide the information or the location of their IntegrationTestScenarios through Git references like
**Github URL**, **Git Options** (**Revision** & **Path in repository**). 

The information provided by the users in the above form will then be consumed by the integration service with the help of IntegrationTestScenario CR and will be used to run the integration tests against the Application Snapshot in question.

Example of the tekton pipeline definition with git resolver:

![](../diagrams/ADR-0023/tekton-pipeline-definition-git-resolver.jpg)

## Consequences

* Better user experience through Tekton pipeline definitions that are hosted in git repositories which are easy to maintain

> with faster execution as opposed to build and maintain the Tekton bundle images.

Reviewer: what faster execution?

Author: Doesn't make sense, removed

* Resolvers are currently in Technology Preview for the OpenShift Pipelines 1.9. Technology Preview features are not
supported with Red Hat production service level agreements (SLAs) and might not be functionally complete.
Red Hat does not recommend using them in production. These features only provide early access to upcoming product features,
enabling customers to test functionality and provide feedback during the development process.

## Footnotes

* The risk mentioned in the consequences has been [accepted] by the PM(s).

> * With the current design, there will be No direct support for the Tekton bundles. 

Author: Was mostly to highlight that part but changed to all smaller

to allow using different types of Tekton resolvers ex: bundles-resolver, cluster-resolver, hub-resolver along with 
the git-resolver. This gives the extensibility to permit the use of different resolvers in the future.

The new format of adding the Integration Test Scenarios would look like:

![](../diagrams/ADR-0023/git-references-ITS.jpg)

Users git repository path to their test code can now provide the information or the location of their IntegrationTestScenarios
through Git references like **Github URL**, **Git Options** (**Revision** & **Path in repository**). 

The information provided by the users in the above form will then be consumed by the integration service with the help of IntegrationTestScenario CR and will be used to run the integration tests against the Application Snapshot in question.

Example of the tekton pipeline definition with git resolver:

![](../diagrams/ADR-0023/tekton-pipeline-definition-git-resolver.jpg)

The existing IntegrationTestScenarios will be migrated to the new API version with tekton resolvers, by using the 
standard kubernetes conversion webhooks ([STONEINTG-386]).

> ## Consequences

Reviewer 3: Can we address the impact of imposter git commits (`https://github.com/tektoncd/pipeline/issues/6315`)? Do we have any plan to help protect (or warn) users against them?

Author: I don't think we have it covered under the ongoing feature [RHTAP-402](https://issues.redhat.com/browse/RHTAP-402). Would you mind filing a jira as a follow-up for the improvements ?

Reviewer 3: Using either open questions as in [ADR 3](https://github.com/redhat-appstudio/book/blob/main/ADR/0003-interacting-with-internal-services.md#open-questions) or out of scope as in [ADR 21](https://github.com/redhat-appstudio/book/blob/main/ADR/0021-partner-tasks.md#out-of-scope) would work for covering in this AR.

Reviewer 4: This looks to me like a Tekton issue that will need to be handled in Tekton and then adopted by the Integration service if there are changes that need to be addressed. It feels like it could go to open questions, but specifying that the Integration service would adopt future security improvements from Tekton as soon as they become available. 
Does that sound good @sonam1412 @arewm?

Reviewer 3: I agree that it is a Tekton issue, but this is also the first time that we are exposing the ability for users to be faced with this type of situation. We should be up front with the fact if we are not doing anything additional to mitigate this risk. Any documentation should also make sure to mention this potential attack vector. 

@sonam1412 shared a proposed wording with me, I assume it just needs to be committed to address this.

Author: Added it [here](https://github.com/redhat-appstudio/book/pull/92/files#diff-4adb99d2e0cdfd21098edf4375e38cd0ca48caeb8f285113121a39634a3989abR51-R56). 
Please take a look.

Reviewer 4: Ack, looks good to me.
